File: src/lib/lseLive.ts
/** London Strategic Edge vault + last-print helpers. Server-side only. */

export const LSE_VAULT_CANDLES = "https://api.londonstrategicedge.com/vault/candles";
export const LSE_WS_URL = "wss://data-ws.londonstrategicedge.com";

export type LseCandle = {
  time: string | number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number;
};

export function lseApiKey(): string {
  return (process.env.LSE_API_KEY ?? "").trim();
}

export function lseStreamConfigured(): boolean {
  return lseApiKey().length > 0;
}

export const LSE_QUOTE_FRESH_MS = 45_000;

export function quoteIsFresh(updatedAt: string | null | undefined, now = Date.now()): boolean {
  if (!updatedAt) return false;
  const t = new Date(updatedAt).getTime();
  return Number.isFinite(t) && now - t < LSE_QUOTE_FRESH_MS;
}

function clampBar(o: number, h: number, l: number, c: number, vol?: number, time?: string | number): LseCandle | null {
  if (![o, h, l, c].every(Number.isFinite)) return null;
  const high = Math.max(h, o, c);
  const low = Math.min(l, o, c);
  if (low > high) return null;
  const t = toChartTime(time);
  if (t == null) return null;
  return { time: t, open: o, high, low, close: c, volume: vol };
}

/** Lightweight-charts: daily `YYYY-MM-DD`, intraday unix seconds. */
export function toChartTime(ts: unknown): string | number | undefined {
  if (typeof ts === "number" && Number.isFinite(ts)) {
    return Math.floor(ts > 1e12 ? ts / 1000 : ts);
  }
  if (typeof ts !== "string") return undefined;
  const s = ts.trim();
  if (/^\d{4}-\d{2}-\d{2}$/.test(s)) return s;
  const iso = s.includes("T") ? s : s.replace(" ", "T");
  const stamped = /Z$|[+-]\d{2}:?\d{2}$/.test(iso) ? iso : `${iso}Z`;
  const ms = Date.parse(stamped);
  if (!Number.isFinite(ms)) return undefined;
  return Math.floor(ms / 1000);
}

export async function fetchLseVaultCandles(
  ticker: string,
  timeframe = "5m",
  limit = 400
): Promise<LseCandle[]> {
  const key = lseApiKey();
  if (!key) return [];
  const url = new URL(LSE_VAULT_CANDLES);
  url.searchParams.set("symbol", ticker);
  url.searchParams.set("timeframe", timeframe);
  url.searchParams.set("limit", String(limit));
  url.searchParams.set("order", "desc");
  const res = await fetch(url, {
    headers: {
      "x-api-key": key,
      "User-Agent": "OrbisEquity-Terminal (+https://londonstrategicedge.com)",
    },
    cache: "no-store",
    signal: AbortSignal.timeout(8000),
  });
  if (!res.ok) return [];
  const body = await res.json();
  const rows = Array.isArray(body) ? body : [];
  const out: LseCandle[] = [];
  for (const r of rows) {
    const ts = r?.timestamp ?? r?.ts ?? r?.t ?? r?.time;
    const bar = clampBar(
      Number(r?.open ?? r?.o),
      Number(r?.high ?? r?.h),
      Number(r?.low ?? r?.l),
      Number(r?.close ?? r?.c),
      r?.volume != null || r?.v != null ? Number(r.volume ?? r.v) : undefined,
      typeof ts === "number" ? ts : ts != null ? String(ts) : undefined
    );
    if (bar) out.push(bar);
  }
  out.sort((a, b) => (a.time < b.time ? -1 : a.time > b.time ? 1 : 0));
  return out;
}

File: pipeline/clients/lse_live.py
"""London Strategic Edge live tape — vault candles + websocket ticks.

Official client is ``lse-data`` (``from lse import LSE``). The old
PostgREST catalog at /iso is batch fundamentals, not a stream.
Live prints come from ``wss://data-ws.londonstrategicedge.com``.
Intraday history comes from ``https://api.londonstrategicedge.com/vault/candles``.

Do not write ticks into ``prices_daily``.
"""
from __future__ import annotations

import os
from datetime import datetime, timezone
from typing import Any, Iterable

VAULT_CANDLES = "https://api.londonstrategicedge.com/vault/candles"
WS_URL = "wss://data-ws.londonstrategicedge.com"
LIVE_TIMEFRAMES = ("1s", "5s", "15s", "30s", "1m", "3m", "5m", "15m", "30m", "1h", "4h")


def lse_api_key() -> str:
    return (os.getenv("LSE_API_KEY") or "").strip()


def lse_stream_configured() -> bool:
    return bool(lse_api_key())


def normalize_symbol(symbol: str) -> str:
    """Keep catalog form (AAPL, BRK.B, BTC/USD)."""
    return (symbol or "").strip().upper()


def tick_to_quote(tick: Any) -> dict | None:
    """Map an ``lse.Tick`` (or tick-shaped dict) to a ``quotes_last`` row."""
    if tick is None:
        return None
    if isinstance(tick, dict):
        symbol = tick.get("symbol")
        price = tick.get("price") or tick.get("last")
        bid = tick.get("bid")
        ask = tick.get("ask")
        volume = tick.get("volume")
        ts = tick.get("timestamp") or tick.get("ts")
        replay = bool(tick.get("replay"))
    else:
        symbol = getattr(tick, "symbol", None)
        price = getattr(tick, "price", None)
        bid = getattr(tick, "bid", None)
        ask = getattr(tick, "ask", None)
        volume = getattr(tick, "volume", None)
        ts = getattr(tick, "timestamp", None)
        replay = bool(getattr(tick, "replay", False))

    try:
        last = float(price)
    except (TypeError, ValueError):
        return None
    if not symbol or last <= 0 or not (last == last):
        return None

    parsed_ts = None
    if ts:
        try:
            parsed_ts = datetime.fromisoformat(str(ts).replace("Z", "+00:00"))
        except ValueError:
            parsed_ts = None

    return {
        "symbol": normalize_symbol(str(symbol)),
        "last": last,
        "bid": float(bid) if bid is not None else None,
        "ask": float(ask) if ask is not None else None,
        "volume": float(volume) if volume is not None else None,
        "ts": (parsed_ts or datetime.now(timezone.utc)).isoformat(),
        "source": "lse_ws",
        "replay": replay,
        "updated_at": datetime.now(timezone.utc).isoformat(),
    }


def vault_candle_bar(row: dict) -> dict | None:
    """Normalize a vault candle row to the chart candle shape."""
    if not row:
        return None
    ts = row.get("timestamp") or row.get("ts") or row.get("t") or row.get("time")
    time_val: str | int | None
    if isinstance(ts, (int, float)) and ts > 1e9:
        time_val = int(ts if ts < 1e12 else ts / 1000)
    elif isinstance(ts, str) and len(ts) >= 10:
        raw = ts.strip()
        if len(raw) == 10 and raw[4] == "-":
            time_val = raw
        else:
            iso = raw.replace(" ", "T", 1)
            if not (iso.endswith("Z") or "+" in iso[10:] or iso.count("-") > 2):
                iso = iso + "Z"
            try:
                time_val = int(datetime.fromisoformat(iso.replace("Z", "+00:00")).timestamp())
            except ValueError:
                return None
    else:
        return None
    try:
        o = float(row.get("open") if row.get("open") is not None else row.get("o"))
        h = float(row.get("high") if row.get("high") is not None else row.get("h"))
        low = float(row.get("low") if row.get("low") is not None else row.get("l"))
        c = float(row.get("close") if row.get("close") is not None else row.get("c"))
    except (TypeError, ValueError):
        return None
    if not all(x == x for x in (o, h, low, c)):
        return None
    high = max(h, o, c)
    lo = min(low, o, c)
    vol = row.get("volume") if row.get("volume") is not None else row.get("v")
    bar = {"time": time_val, "open": o, "high": high, "low": lo, "close": c}
    if vol is not None:
        try:
            bar["volume"] = float(vol)
        except (TypeError, ValueError):
            pass
    return bar


def make_client(api_key: str | None = None):
    from lse import LSE

    return LSE(api_key=api_key or lse_api_key())


def fetch_vault_candles(symbol: str, timeframe: str = "5m", limit: int = 400) -> list[dict]:
    """REST vault candles. Requires ``LSE_API_KEY``."""
    client = make_client()
    rows = client.candles(normalize_symbol(symbol), timeframe, limit=limit, order="desc")
    bars = [b for b in (vault_candle_bar(r) for r in rows) if b]
    bars.sort(key=lambda b: b["time"])
    return bars


def iter_ticks(symbols: Iterable[str], *, replay_start: str | None = None):
    """Blocking iterator of live ticks from the official LSE websocket."""
    client = make_client()
    names = [normalize_symbol(s) for s in symbols if s]
    if not names:
        return
    for tick in client.stream(names, reconnect=True, start=replay_start):
        yield tick
